### PR TITLE
CB-10092: Adding an internal loadbalancer to YCloud datalake setups.

### PR DESCRIPTION
**Pull request by sxxgrc**
Jira: https://jira.cloudera.com/browse/CB-10092

### Status

This has been tested fully, end-to-end, with additional verification that nothing has been broken for AWS/Azure LD/HA setups.

### Main Changes

- A launchLoadBalancer() method was added to the stack creation flow within the launchStack logic
- A new class called YarnLaunchLoadBalancerService was added which handles all of the launch logic
- The gateway nodes created before the launch loadbalancer steps have their IPs retrieved
- A new docker image (which will be hosted where the default YCloud image is hosted) was created and is being pointed to by the LoadBalancer container created, this image will install the haproxy service and point it at the provided servers
- The loadbalancer container's IP is retrieved in the same way as other YCloud container IPs, but we take advantage of the fact that we know what the loadbalancer application name is and use that instead of anything more complex

### CODE REVIEW

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java
Comment (brycederriso) [inline cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java:107]: **question:** Why are we adding this `launchLoadBalancer` method? I suppose it makes sense conceptually to include a separate API handle for it, but we generally don't have that for other "parts" of the cloud stack.

The `launch` method just handles the whole `CloudStack` object, of which the load balancers are a part.

Comment (brycederriso) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java:107]: I can see the utility of something like this when we implement an _upgrade_ path for LBs though. @hreeve-cloudera probably has better ideas here than I do.

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/ApplicationNameUtil.java
Comment (brycederriso) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/ApplicationNameUtil.java:46]: **nit:** Use assignment and plus rather than the `+=` operator.
```suggestion
            name = name + "-" + userName;
```

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnMetadataCollector.java
Comment (brycederriso) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnMetadataCollector.java:153]: **question:** I don't think I'm following this method, will we always have the same number of Containers as `LoadBalancerType`s?

Here, the metadata is responsible for _assigning_ rather than looking up the container information?

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnResourceConnector.java
Comment (brycederriso) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnResourceConnector.java:126]: **praise:** Good method extraction.

**Review by brycederriso — COMMENTED** [review]
Left some comments, but didn't really get to dig in to `LoadBalancerLaunchService`.

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java
Comment (sxxgrc) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java:107]: At least for the Yarn case this was kind of necessary to get all of the logic in the right place. Essentially, we wanted to be able to launch the loadbalancer after the other containers were launched so we could get their IP addresses, but still have the logic in a place where we had access to the stack and could persists the resource we would create for the LB, and this was the best place for it.
Haley mentioned that this would be better in order to isolate the logic completely to help with future changes and the like.

File: cloud-api/src/main/java/com/sequenceiq/cloudbreak/cloud/ResourceConnector.java
Comment (sxxgrc) [reply cloud-api/src/main/java/com/sequenceiq/cloudbreak/cloud/ResourceConnector.java:66]: Will do, thank you.

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/ApplicationNameUtil.java
Comment (sxxgrc) [reply cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/ApplicationNameUtil.java:46]: Will do!

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnMetadataCollector.java
Comment (sxxgrc) [reply cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnMetadataCollector.java:153]: Good question, for Yarn this will always be true since Yarn will only ever have public loadbalancers.
In general though, as this method is for getting all of the loadbalancer metadata, and the only inpus is a list of the types, I kind of assume that at most there will be the same number of loadbalancers as the amount of unique types used. @hreeve-cloudera can clarify this some more.

I was just trying to make sure I future proof this in case Yarn does end up using private loadbalancers and the like in the future.

**Comment by sxxgrc** [comment]
Made @brycederriso 's desired changes and answered questions.

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java
Comment (hreeve-cloudera) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java:107]: We do have an upgrade path for load balancers. Bryce, you and I have talked about it briefly. But basically in CB-10612 I added a new API that, if called, will add load balancers to an existing environment. As part of that I separated out the AWS load balancer logic into a new `AwsLoadBalancerLaunchService`. So for AWS rolling with this change will be simple, because we can simply move where `AwsLoadBalancerLaunchService` is being currently called to be called here instead.

Ultimately we need to be able to call the LB creation logic separately of the stack creation anyway, to support that in-place upgrade API. So I didn't see any harm in separating things out like that, and told Santiago to go for it, and we would figure out how to implement it across the other cloud platforms as necessary.

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java
Comment (hreeve-cloudera) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java:133]: Since `LoadBalancerLaunchService` is an injectable object, don't make this static. Instead just inject `LoadBalancerLaunchService` where this needs to be called, and call it through that object.

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnMetadataCollector.java
Comment (hreeve-cloudera) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnMetadataCollector.java:156]: Where is this naming convention being set? If it's used in other places, then we should have a shared method that can be called to generate this, so that it doesn't get out of sync if we change it one place and forget to update it elsewhere.

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java
Comment (hreeve-cloudera) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java:50]: Could you name this `YarnLoadBalancerLaunchService`?

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java
Comment (hreeve-cloudera) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java:181]: Is this the other place you're setting the LB name? Referring to my comment in `YarnMetadataCollector.collectLoadBalancer`. Yeah, please make some shared method both of these can call to generate the LB name. I realize it's just a simple string concatenation, but we don't want these to end up mismatched.

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnMetadataCollector.java
Comment (hreeve-cloudera) [reply cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnMetadataCollector.java:153]: It looks to me like in `LoadBalancerLaunchService`, you're creating one component per load balancer in the CloudStack, and each component has one container. So logically if there were two load balancers, there would be two components with a combined two containers. Is that right? That makes logical sense to me, and would match what seems to be being done in the for loop here.

File: cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java
Comment (hreeve-cloudera) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java:107]: Actually, Santiago, I don't think you need a new method here. This is functionally just a duplicate of the `updateLoadBalancers` method I already added to the `ResourceConnector` interface. So let's consolidate that. (We can rename the existing method to `launchLoadBalancer` if you like, to match the other `launch` method.)

Comment (sxxgrc) [reply cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/connector/resource/AwsResourceConnector.java:107]: When in the flow is that called? It might require a little bit of work to get it working in that order, but not sure.

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java
Comment (sxxgrc) [reply cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java:133]: Ahh that makes sense, sorry still getting used to Spring stuff.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (sxxgrc) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:193]: Yes that is true.

File: core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java
Comment (sxxgrc) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:254]: Whoops sorry, I remember you had said in my previous PR for removing the entitlement that this order was more optimal since it just checks a string first, and then checks the entitlement which is a more expensive operation.
I can revert this if you want.

Comment (hreeve-cloudera) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/LoadBalancerConfigService.java:254]: I suspected that may be the reason :). That is generally true, yes, but in this particular case I did it for a reason. You couldn't have known that though, so I appreciate you trying to optimize the code. But yes, please put it back.

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java
Comment (sxxgrc) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:75]: Will do before I merge.

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java
Comment (sxxgrc) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:120]: So this method is in another package than the one for the LaunchStackHandler, which would require a dependency to be added which tbh I have not done before. How should I handle that?

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java
Comment (hreeve-cloudera) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:75]: Ah, so it isn't. Thinking on it more, I think you can just check if `cloudStack.getLoadBalancers().isEmpty()`. The load balancers objects are only created if all the validation checks in `LoadBalancerConfigService` passed, so I don't think we need to check them again.

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java
Comment (hreeve-cloudera) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:120]: Ah, so it isn't. Thinking on it more, I think you can just check if `cloudStack.getLoadBalancers().isEmpty()`. The load balancers objects are only created if all the validation checks in `LoadBalancerConfigService` passed, so I don't think we need to check them again.

Comment (sxxgrc) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:120]: Okay I'll do that then.

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnResourceConnector.java
Comment (biharitomi) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/YarnResourceConnector.java:205]: Can we have this on INFO level?

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java
Comment (biharitomi) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java:82]: Can we have INFO level here?

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java
Comment (biharitomi) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java:125]: Info level?

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java
Comment (biharitomi) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java:143]: Warn level?

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java
Comment (biharitomi) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java:152]: Warn level?

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java
Comment (biharitomi) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java:148]: Info level?

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java
Comment (biharitomi) [inline cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/YarnLoadBalancerLaunchService.java:174]: I would move the creation of the components into an individual service/factory. The `YarnLoadBalancerLaunchService.java` would be much more clean.

**Review by biharitomi — CHANGES_REQUESTED** [review]

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java
Comment (sxxgrc) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:68]: What exactly do you mean by this? Should I check to see if the status is equal to FAILED or DELETED?
And if so, should I just not do the load balancer stuff, or handle that differently?

**Comment by sxxgrc** [comment]
Made most of the changes requested by @biharitomi 
I went ahead and resolved most of the conversations here as well to make it easier to go through this PR. I left only the ones I would like to be double checked as well as the single requested change from Tamas that I want further clarification on before I work on.

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java
Comment (hreeve-cloudera) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:68]: @sxxgrc The `ResourcesStatePollerResult` contains the state of all the resources created in the cloud. At this point in the process, that should be all the instances, network resources, etc. What Tamas means is that we should check the state of those resources, and if they're failed we shouldn't do the load balancer stuff, since the creation will ultimately fail anyway. I think that would look something like this:
```
ResourcesStatePollerResult statePollerResult = waitForResources(ac, resourceStatus, cloudContext);
if (statePollerResult.getResults().stream().noneMatch(result -> ResourceStatus.FAILED.equals(result.getStatus()))) {
    launchLoadBalancers(cloudStack, connector, ac, persistenceNotifier, cloudContext, statePollerResult);
} else {
    LOGGER.error("Some cloud resources failed creation. Load balancers will not be created.");
}
LaunchStackResult result = ResourcesStatePollerResults.transformToLaunchStackResult(request, statePollerResult);
```
This isn't really something you've introduced. It should have been done earlier, but wasn't.

Comment (biharitomi) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:68]: @sxxgrc as @hreeve-cloudera mentioned can we get that additional condition? I can imagine that condition in the new `launchLoadBalancers` method too.

**Comment by biharitomi** [comment]
I checked the code after you have processed the review items and only one more review item left:
https://github.com/hortonworks/cloudbreak/pull/10230/files#r599465559
Otherwise it looks good to me.

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java
Comment (sxxgrc) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:69]: Okay I added this condition to the launchLoadBalancers method. Let me know if y'all need anything else.

**Comment by sxxgrc** [comment]
Made the last change requested by @biharitomi

File: cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java
Comment (sxxgrc) [reply cloud-reactor/src/main/java/com/sequenceiq/cloudbreak/cloud/handler/LaunchStackHandler.java:69]: By the way, the resulting LaunchStackResult object will contain the statuses for both the initial datalake instance launch and the loadbalancer launch, so that it will be aware of any failures across both launches and that will be handled by the general launch failure handling done later in the flow.

**Comment by sxxgrc** [comment]
Added the final docker image endpoint, which can be found here: 
https://github.com/hortonworks/cloudbreak/pull/10230/files#diff-4ed62beae7d4bfe342a477f7137505bb4f17d6860d5043cd505353369ef5580dR37

I will run some final tests on this, but I will make this mergeable now.

File: cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java
Comment (sxxgrc) [reply cloud-yarn/src/main/java/com/sequenceiq/cloudbreak/cloud/yarn/loadbalancer/service/launch/LoadBalancerLaunchService.java]: I created the `YarnApplicationCreationService` class to remedy this, putting all of the methods they both call there. Now only the `YarnResourceConnector` injects the `LoadBalancerLaunchService` but not vice versa.

**Review by hreeve-cloudera — APPROVED** [review]

**Review by biharitomi — APPROVED** [review]
LGTM

**Pull request merged**
